### ISSUE TITLE
BrowserLikePolicyForHTTPS always ignores the specified trustRoot and uses the system trust root instead

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7647]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7647|
|Type|release blocker: regression|
|Created|2014-09-17 10:34:27Z|
|Branch|https://github.com/twisted/twisted/tree/browserlikepolicy-secfix-7647|

<details><summary>Searchable metadata</summary>

```
trac-id__7647 7647
type__release_blocker__regression release blocker: regression
reporter__hawkowl hawkowl
priority__highest highest
milestone__None None
branch__branches_browserlikepolicy_secfix_7647 branches/browserlikepolicy-secfix-7647
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__web web
keywords__security security
time__1410950067628664 1410950067628664
changetime__1445217216921098 1445217216921098
version__None None
owner__glyph glyph
cc__jknight cc__hynek
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

(In [43122]) Branching to browserlikepolicy-secfix-7647.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)|@hawkowl set owner to @glyph|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

Thanks for your work on this.

In `twisted/web/test/test_agent.py`:

> from twisted.internet._sslverify import ClientTLSOptions

So... abstraction boundary violation.  Things are private for a reason, right?  Why does a Twisted Web unit test have to depend on private APIs from `twisted.internet`?  Of course, this happened already...  This branch just makes it worse.

So do I let it slide because the basic problem is already in place and this is blocking a security fix?  Maybe not, because a security release was already made.  We can take the time to fix this right in trunk.

It's still really difficult to understand how all of these pieces - some private in `twisted.internet._sslverify` and OpenSSL-specific, some public in `twisted.internet.interfaces` and OpenSSL-specific, some private but apparently required to test "third party" (`twisted.web.client`) applications that _use_ the public pieces - are intended to fit together.  I'd really like to see some documentation about these things.  Then maybe it'll be possible to do a good review of changes like this one.

Lacking that, all I can say is that `twisted.internet.ssl` should somehow offer some public APIs to help out writing unit tests and `twisted.web.test.test_agent` should use those instead of internal implementation details.

Thanks again.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_6"></a>|@hawkowl commented|
|-|-|

Putting this on 14.1, because unless we backport (front-port?) the 14.0.2 fix, it'll release with a security issue.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_8"></a>|@glyph commented|
|-|-|

Pascal's apology applies, multiple times, to the following.  You might want to Instapaper it or something, get a limited print run from lulu, maybe sell some ebooks of it to offset the cost of the time you had to spend reading it. Sorry.

This is at least 2 gigantic essays: one on workflow and one on the architecture of TLS within Twisted.

Replying to [exarkun](#note_5):
> Thanks for your work on this.

Thanks for reviewing.

> In `twisted/web/test/test_agent.py`:

> > from twisted.internet._sslverify import ClientTLSOptions

> So... abstraction boundary violation.

I've spent many days trying to come up with a _fix_ for these comments, I guess I might as well try to just answer each of your questions with all of the thoughts I've had in the interim.  Perhaps a fix will emerge but perhaps not.

> Things are private for a reason, right?

Perhaps this is worth exploring, first.  Things are private for many reasons.

The first and most important in Twisted's case is to preserve future flexibility and allow us to make progress without an overwhelming cycle of deprecations for users.  While our deprecation tooling is improving, it still leaves a lot to be desired, and even with perfect tooling, dealing with deprecations and removals is a hassle.

A secondary reason that things are private is that more public things means more public interfaces which means more things for developers to learn.

"developers" then breaks down into two audiences: external developers and internal maintainers.

In this case, I think that the effort saved, for now, in breaking the internal abstraction boundary, more than justifies the cost of imposing a little cognitive burden on maintainers, especially since the cognitive burden related to SSL is already so crushing that it's hard to make it worse.

> Why does a Twisted Web unit test have to depend on private APIs from `twisted.internet`?  Of course, this happened already...  This branch just makes it worse.

The reason that this test needs to depend on private `twisted.internet` APIs is that the information we really want in the test - what trust roots are specified to the client created by the `Agent` - is buried inside OpenSSL and already unavailable for inspection. By "already" I mean "already in Twisted core, before Twisted web gets involved".  Therefore `twisted.internet` doesn't have any useful object to expose.  There's no "good" way to write it, at least none that I've figured out yet.

Getting down into the nitty-gritty specifics, `OpenSSL.crypto.X509StoreType` has exactly one public method: `add_cert`.  Presumably, `OpenSSL.SSL.ContextType.set_default_verify_paths` manipulates this thing's internal state in some way, but we can't observe that.

So, we can't expose the `X509StoreType` directly, because it's useless.  We also can't expose a wrapper around the whole `Context` mess that reports on trust roots, because we can't introspect `X509StoreType` sufficiently to determine what `set_default_verify_paths` even did to it; so any wrapper would need to consist of some `twisted.internet.ssl.Certificate` objects that we _can_ see, and then some `twisted.internet.ssl.HeckIfIKnow` objects which we can't.  A combination of objects with methods and opaque handles to unknown things seems like a pretty bad return type for a method, so I left it out.

> So do I let it slide because the basic problem is already in place and this is blocking a security fix?  Maybe not, because a security release was already made.  We can take the time to fix this right in trunk.

I think that we should adopt something like the ["yes, and" rule](https://en.wikipedia.org/wiki/%22Yes%2C_And%22_rule): generally speaking we should land things of acceptable quality on trunk and then build upon them with further tickets.

By this I don't mean that we should accept modifications which fall short of our standards for quality, but that we should have a certain standard for quality and then we should stick to that, and not impose additional burdens.

I've been advocating this approach softly for quite a while now, but I don't know if I've been very clear in my reasoning, and this ticket provides an excellent case-study.

This review feedback proposes three new-to-me requirements for Twisted code:

    1. we need to avoid all internal private API usage, even in test modules

    2. changes which interact with subtle interactions of dependent subsystems should come along with high-level architectural documentation of all the relevant details of those subsystems

    3. changes which establish a need for a test mock within Twisted must come along with an published, documented, tested verified fake of that same mock

I think it would be wonderful to _have_ all of those things.  I'm very glad that a reviewer noticed them and brought them to my attention.  But, I don't think that introducing scope-creep to a bug-fix (especially a release-blocking security bug-fix) is the right way to get them.

Moreover, feedback that introduces new requirements like this is incredibly discouraging to me personally for several reasons.  First, it increases the scope of work necessary to get a fix approved and landed, which means the feeling of accomplishment associated with getting a ticket closed is deferred yet further.  It also increases the meta-scope: we tend to elevate desirable qualities to categorical imperatives, and for good reason: if those attributes are _required_ here, why shouldn't they be required for other changes?

Second, once I have feedback like this, it feel like "cheating" to solicit further reviews of subsequent changes by other people, because since you’ve felt it necessary to step outside the documented requirements for a branch to be landed on trunk, getting a review from someone else who might not have the same extraordinary concerns seems like a cheap route to get it landed.  So if nobody else is reviewing it I feel more obliged than usual to wait.

Third, it interacts with another community antipattern we have, which is persistent complaining about change size.  Many reviewers vocally complain on a regular basis that changes are too big, and that big changes are too hard to review.  It is also difficult both in terms of verison control tooling and in terms of workflow to take a large existing change and split it up, which means there is more pressure to just get small changes landed so I can build upon them.

Basically, it's exhausting to work on Twisted because the goal posts move.  And they don't just move; the effect of the tension between "all new and changed code must be conceptually coherent, feature-complete, fully documented at the API level, at the narrative level, and at the architectural level" and "all changes must be small incremental changes which can be easily reviewed" means that the left goal-post is actually to the right of the right goal-post; for some features which require both concision and global architectural improvement, the goal is of negative width.

I wouldn't presume to speak authoritatively to anyone else's source of motivation but I would definitely hazard a guess that this pattern is discouraging to others as well.

[ticket:2673#comment28 You mentioned on another ticket] that “‘ticket not approved’ is pretty much the only leverage ![you] have”.  As it turns out, unfortunately for both of us, you don’t really have _any_ “leverage”, and this probably isn’t the best way to get people to do more work.  (And I really do mean "both of us", I sure do wish you could just push this button and get me to pop out more amazing work ;-)).

I have professional time to devote to Twisted, but even in my case, two factors confound the “ticket not approved as leverage” approach: first, I have other professional responsibilities I can easily attend to instead when a ticket gets “stuck” and I’m not sure how to proceed, and Twisted itself is big enough and has enough of a (review/ticket/backlog) that even if I'm going to work all day, every day, on Twisted, I have more than enough fodder for other activities.  So I suspect this problem is made even worse for others.

My goal when we originally established the [review process](https://twistedmatrix.com/trac/wiki/ReviewProcess) was not to figure out a way to squeeze more work out of every contribution, but to make sure that each contribution was a _net improvement to Twisted’s quality_.

It's still critically important to require documentation and tests for every change, because I still think the goal of preventing regressions is important. I don't want to look at this as a trade-off between sacrificing our agreed-upon standards to "get things done in some cases.  But our agreed-upon standards must be finite, bounded, and fair.

The secondary function of the review process to discover and illuminate _related_ issues is also incredibly useful, but there really needs to be a _very_ sharp division between blocking comments, "consider this" comments, and "please file a new ticket" comments.

Finally - and here I don't need to speculate about other people's motivation, because I've discussed this directly with several reviewers in the past - this style of review from a Quite Senior Indeed twisted reviewer tends to make newer contributors think there is a deep hidden mystery behind doing good reviews. Since none of the stuff you raised here is really required by any of the relevant documentation on how to do reviews, and yet it all appears to be blocking feedback, this leads newer reviewers to ahve significant anxiety they will Miss Something, which would of course be terrible, so they don't even try (and so they don't build their skills, and so on).  Worse than setting a bad example, it convinces people that they cannot possibly succeed because the criteria for success require secret knowledge.

> It's still really difficult to understand how all of these pieces - some private in `twisted.internet._sslverify` and OpenSSL-specific, some public in `twisted.internet.interfaces` and OpenSSL-specific, some private but apparently required to test "third party" (`twisted.web.client`) applications that _use_ the public pieces - are intended to fit together.  I'd really like to see some documentation about these things.

One of my other difficulties in responding to this feedback is that I don't know where this documentation should go.  This is not to say that I don't think it should be written - clearly my architectural vision for what `twisted.internet.ssl` is has been _extremely_ poorly communicated, and that's a problem.  Our "good" new SSL API, `twisted.internet._sslverify.OpenSSLCertificateOptions`, was a class I created that I never even intended to be public.  Its name derives from the fact that it is the return value of `PrivateCertificate.options`.  Its construction signature has evolved into a hybrid mishmash of "low level" PyOpenSSL-specific junk and "high level" objects like the `Certificate` wrapper from Twisted itself.  It is re-exported as `twisted.internet.ssl.CertificateOptions` despite the fact that its whole API structure is deeply, deeply OpenSSL specific. (Which makes sense as it was originally designed as the OpenSSL-specific implementation detail behind the higher-level `options` method!)

Nevertheless, it would be wrong to say that `PrivateCertificate.options` is the _correct_ API to use today, because it is missing several critical features which have been added to `OpenSSLCertificateOptions`.

This apparently contradicts what I said earlier about it being OK to impose more of a cognitive burden on people working with SSL because it's so hard already, because that strategy has clearly resulted in a completely dysfunctional architecture in this code.  Basically, I believe we're at a local maximum here, and without some kind of radical change we are not going to be able to make substantial whole-system progress; this change is just maintanence and incremental improvement of the status quo.

So let me try to answer your question about how _I_ think this whole mess is supposed to fit together, all the way from `BrowserLikePolicyForHTTPS` down to `CertificateOptions` and eventually `transport.startTLS`.  Maybe with that information in hand, you can tell me where a reasonable place for it would be. First, let me give you a functional description of what actually happens when you make an HTTPS request with Agent.

HTTPS provides some information from a potential request - specifically, hostname and port number - that is relevant to TLS.  In Twisted's current model, the request provides the host name and port number, and the Agent has a way of converting that into a connectionCreator (née "context factory"; we'll get to that in a minute).

`Agent` has an internal separation of concerns with `self._endpointFactory`. When it needs to create a client connection, it does `self._endpointFactory.endpointForURI` to create an endpoint.  Using `Agent.usingEndpointFactory`, applications can do an end-run around this whole dance entirely by providing their own endpoint, but `_StandardEndpointFactory.endpointForURI` honors the `_policyForHTTPS` - a provider of `IPolicyForHTTPS` - that it was constructed with, which is the one normally passed to `Agent`, and constructs an `SSL4Endpoint`.

So, for an "`https`" scheme URI, `_StandardEndpointFactory.endpointForURI` calls `_policyForHTTPS.creatorForNetloc(host, port)`.  It then passes the result of that - the connectionCreator/contextFactory - straight to `SSL4ClientEndpoint`, which passes it to `reactor.connectSSL`, which passes it to `TLSMemoryBIOFactory`.

Now, when `TLSMemoryBIOFactory` is constructed, it has to choose two things: whether it's a client or a server, and whether it's "new-style" (`connectionCreator`, `IOpenSSL______ConnectionCreator`) or old-style (`contextFactory`, `twisted.internet.ssl.ContextFactory`).  It detects this by looking at whether its argument provides the appropriate interface for its client-ness or server-ness, `IOpenSSLClientConnectionCreator` for clients and `IOpenSSLServerConnectionCreator` for servers.

The difference in behavior between "new style" and "old style" has to do with whether the Protocol can depend on the `app_data` on the connection and `info_callback` on the context object being left alone or not.  With new-style stuff, the app_data _might_ be used by Twisted to store a backreference from the TLS protocol from the TLS connection object so that the negotiation can inform the protocol.  The signature has been changed to facilitate this, so that the `connectionCreator` owns the entire process of creating an OpenSSL `Connection` object, as opposed to the `Context` coming from the `ContextFactory` and `twisted.protocols.tls` doing the other half where it actually builds the `Connection`.  (The only thing that makes good use of this new protocol at this point is `optionsForClientTLS`.)

If you're in "old-style" mode, there's simply an adapter from the old to the new style, `_ContextFactoryToConnectionFactory`, which `TLSMemoryBIOFactory` wraps its argument with if necessary.  So past this point, when the `TLSMemoryBIOFactory` is constructed, everything appears "new style".

Old-style should obviously be deprecated but there are no deprecations yet because this language heavily infests the documentation and examples and we really did not want to discourage anyone from using TLS at all, so we were pretty conservative with making everything work as much like it previously did as possible.

So, where we left off with the new/old difference, we were about to start a TLS connection.  `TLSMemoryBIOProtocol.makeConnection` calls `TLSMemoryBIOFactory._createConnection`, which returns the pyOpenSSL `Connection` object to use.  That in turn calls either `clientConnectionForTLS` (for clients) or `serverConnectionForTLS`; the distinct methods are so that a connection creator can be used for both and know what type of connection it's being asked to create.  They are different methods rather than a flag because the security properties which make sense for TLS clients differ vastly than those for servers, and we wanted it to be obvious to make the distinction.

Hopefully everything from the handoff here into transport-land makes sense, since I'm pretty sure you wrote that part ;-). We have a `Connection`, we do BIO stuff to it, the end.

But we need to go back to what `_StandardEndpointFactory.endpointForURI` actually gets back from `creatorForNetloc`.  Unless the application has supplied its own, that will call `BrowserLikePolicyForHTTPS.creatorForNetloc`, which ignores the port number, decodes its hostname, passes that to `optionsForClientTLS` (and then possibly passes a trust root along as well, on 14.0.2 and if this branch lands).

`optionsForClientTLS` (a public, well-advertised function) takes 4 arguments: a hostname, unicode, which is used for SNI and hostname verification; `trustRoot`, an indication of what roots to trust, which defaults to "platform" trust roots (I'll describe its type momentarily); `clientCertificate`, a `PrivateCertificate` that is used as a client certificate, `extraCertificateOptions`, which is "junk that gets passed through to `CertificateOptions` in case we forgot something".  It returns an `IOpenSSLClientConnectionCreator`.

`optionsForClientTLS`, constructs a (private type), `ClientTLSOptions`, which sets up the `info_callback` on its `Context`, and the `app_data` on each `Connection` it constructs to the protocol for, so that its verification callback can tell the `TLSMemoryBIOProtocol` if verification fails and it can shut down.  To give the `ClientTLSOptions` a `Context` to work on, it uses `CertificateOptions`, and just builds its context once at construction time.

Now, what about `trustRoot`'s type?  It's a provider of the _private_ interface `IOpenSSLTrustRoot`.

This interface is private because the way that `OpenSSL` deals with trust roots is really, really dumb, and I was hoping to avoid exposing that to the user. In contrast to, for example `IOpenSSLClientConnectionCreator`, which can use vaguely sensible pyOpenSSL APIs to construct a TLS connection if you know what you're doing, the way trust roots are dealt with is full of potential platform-specific junk and it's not clear that we've figured out the best strategy.  So the idea is that you get an opaque object from one of two places: `twisted.internet.ssl.platformTrust`, or you can pass a `twisted.internet.ssl.Certificate` which will be used as a CA certificate.  We don't deal with as much of that as I'd hoped to in the initial design for this, but the current specification would allow for a great deal of those smarts to be added without disturbing applications.

The ultimate consumer of `IOpenSSLTrustRoot`, which is to say, `OpenSSLCertificateOptions`, is currently missing a very important third thing, which is a way to combine trust roots.  You can't currently specify multiple manual CAs, or a combination of the platform trust root plus some other CAs.  I am too tired from all this typing at this point to look up whether there's already a ticket for fixing that.  There should be.

This private interface is simply the method `_addCACertsToContext`, which is passed a `Context` (another reason that this is private: it is absolute nonsense that the method for adding platform trust roots to a trust store takes a context and not, you know, the trust store; after going to some trouble to get application code, that is, code outside `twisted.internet.ssl` itself, to manipulate the `Context` objects produced by Twisted, I didn't want to add yet another place where it was passed through).

Finally, how do I intend for this code to evolve, particularly in the hope that one day we might support multiple TLS backends and wean ourselves of our crippling pyOpenSSL addiction?

Right now, the whole implementation is quite OpenSSL-specific.  It is likely to remain so for some time, so creating a completely abstract indirected interface is not currently a goal.  However, I would like to start factoring responsibility into the appropriate places and drawing a line between what Twisted is doing and what it expects the lower-level implementation to do.

My hope is that you would be able to work entirely in terms of high-level concepts in `twisted.internet.ssl`.  For example: `Certificate` should have a wrapper around the `X509` interface in OpenSSL that is complete enough for any Twisted application to get the relevant information out of it.  Currently, implementation details leak through all over the place, like the `format` argument to the `Certificate.load` classmethod, which must be a constant out of pyOpenSSL.  `CertificateOptions` will need to be redesigned entirely, but that public name is so terrible that I'm happy to deprecate it once we have something better in place.  (For that matter `options` is not a great name for the method.)

But enough with the minutae of the interfaces' defects and my hopes for the future.  How exactly do I envision the reactor core and the TLS layer interacting?

It's important to remember the fact that the reactor has to interact with TLS at all is a design error.  So eventually we should be able to get rid of `connectSSL` and `listenSSL` and just get to where we have a `TLSEndpoint` that wraps another endpoint.  `startTLS` is a little trickier but even without going into the perpetual rathole of in-reactor protocol switching, it is not that hard to shunt all the bytes from one protocol into another one.  (And Tubes may one day address that problem, but nobody hold their breath, obviously.)

So ultimately all of this boils down to `TLSMemoryBIOProtocol`.  How does it communicate with this TLS layer?  Well... it doesn't, really.  By its nature, it calls a ton of OpenSSL-specific APIs.  Even its name has "BIO" in it.

That makes the last integration point before we start descending into OpenSSL hell `TLSEndpoint`, and that is indeed where I forsee some kind of implementation-selection occurring.

TLS implementations ought to be able to provide the following entry-points:

    1. `optionsForClientTLS` - more or less like it is now, but perhaps with a few more options extracted from `extraCertificateOptions`, like min/max TLS protocol versions

    2. `optionsForServerTLS` - replacing the role that `CertificateOptions` currently plays; although this should take a `PrivateCertificate` it would probably be a mistake to go back to the strategy of trying to make it a method there, because that would disallow certificateless operation modes like TLS-SRP

    3. `Certificate.load` - or a better version of it that doesn't take an OpenSSL constant

    4. `PublicKey.load` - ditto

    5. `KeyPair.load` - ditto

    6. a new constructor for `PrivateCertificate` that includes chain certificates, and composes out of `Certificate` and `KeyPair`

    7. `platformTrust` - unfortunately we probably need to deal with `OpenSSL`'s crappy trust managment API so we probalby can't make this return anything useful in other implementations either to avoid depending on it.

    8. for completeness, but rarely interesting to applications: `AcceptableCiphers`, `DiffieHellmanParameters`, `DistinguishedName`, `CertificateRequest`.

The return types for `platformTrust`, `optionsForClientTLS`, and `optionsForServerTLS` can all be relatively opaque, at least in the sense that `platformTrust` is only consumed by those two APIs, and `optionsFor______TLS` can be things where `TLSEndpoint` can do something backend-specific.  Perhaps the method they will eventually have will be simply something like `createWrapperProtocolFactory`; they already have all the TLS-related information that they need.  The point is that application code passing an `optionsForClientTLS` to a `TLSEndpoint` (or, depending on how long they live, its legacy aliases, `connectSSL` and `startTLS`, which simply construct the same thing) should not need to change.  The specification of the arguments to `optionsForClientTLS` should be, modulo the `extraCertificateOptions` escape-hatch, sufficiently generic for any TLS backend to handle.

> Then maybe it'll be possible to do a good review of changes like this one.

While I understand and sympathize with your frustration here about the whole TLS subsystem being quite complex and hard to understand, I think changes like _this_ one should be eminently reviewable without a terribly sophisticated high-level view of the whole thing.

The goal of this change is to do exactly one thing: previously, the `trustRoot`, a value specifying the certificate authorities to trust for the purposes of X509 certificate verification, was not passed from `BrowserLikePolicyForHTTPS` (the HTTPS-configuration layer, that knows about URLs (specifically "netloc"s since that's all that should be relevant)) to `optionsForClientTLS` (the protocol-agnostic TLS configuration layer), meaning that HTTPS _always_ got the default trust roots.

This much stuff should all be comprehensible from the existing API documentation, specifically the docstrings for `IPolicyForHTTPS` and `IOpenSSLClientConnectionCreator` and `optionsForClientTLS`, which should all be within arms reach for a reviewer of this change.

There are two things I want to make absolutely sure that the reviewer of this change understands:

  1. why not passing this was so very bad
  2. that the new test properly verifies that the value was in fact passed

Are these things in fact obvious, or am I missing something about how this was communicated?  Is more information (or even just the info-dump I posted above) necessary to understand these things?

The reason I am concerned about this point of feedback is that understanding all of the _internal_ factoring of this multi-layered system is quite complex, and if having to fully understand _all_ of it is a prerequisite for being able to review code which modifies it, then we're going to have an even smaller crowd of reviewers than usual on this system.  I would like to figure out a way to answer your questions so that future reviewers can get a much more targetted explanation.

> Lacking that, all I can say is that `twisted.internet.ssl` should somehow offer some public APIs to help out writing unit tests and `twisted.web.test.test_agent` should use those instead of internal implementation details.

It definitely should, but as I said with far more words above: I think that that is out of scope for this ticket - even after writing all that stuff above, I'm _still_ not sure what that should look like, largely due to deficiencies in OpenSSL :-).

> Thanks again.

As long as I said so much stuff about motivation above, I should note: it does always help, every time, to see this.

Thank you, too.

-glyph

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

So... after reading through this, and being antsy for 14.1, I think I have to ask the question:

Do we release 14.1 with this change, and work on trunk later, doing it properly (if such a thing can be done?)
Do we accept that it's not a perfect solution, and that it works, merge to trunk, and we make a new ticket to improve it later, along with the things that need to be done to improve it?

I honestly have not done anything on this because I do not understand what a solution would even look like, and it seems that a proper solution that does not leak OpenSSL-goop everywhere is both outside the scope of this ticket, and more something that is a long-standing problem, rather than something with this particular change.

I'm not sure. I feel like the first option is harmful because we're applying special patches that aren't in trunk, and the second is harmful because we're making the leaky-abstraction problem worse.

Should I make an Executive Decision(TM) and go for the former, and get *SOMETHING* out the door?  I'm not sure.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

I hope I'll be able to respond/react/whatever to the rest of that significant comment later.  For now I thought it was worth responding to one part of it thought.

> In this case, I think that the effort saved, for now, in breaking the internal abstraction boundary, more than justifies the cost of imposing a little cognitive burden on maintainers, especially since the cognitive burden related to SSL is already so crushing that it's hard to make it worse.

I'm just at the end of a seven hour session of looking at these APIs and the related documentation and tickets.  I feel I'm gradually coming to understand what's going on but I still don't grasp the big picture (and I fear I still have a long way to go).  Whatever else you might want to say about this code, I don't think it's correct to say that what's going on here is "imposing a little cognitive burden".  It's imposing a lot more than that.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_12"></a>|@glyph commented|
|-|-|

Replying to [exarkun](#note_11):
> I hope I'll be able to respond/react/whatever to the rest of that significant comment later.  For now I thought it was worth responding to one part of it thought.
> 
> > In this case, I think that the effort saved, for now, in breaking the internal abstraction boundary, more than justifies the cost of imposing a little cognitive burden on maintainers, especially since the cognitive burden related to SSL is already so crushing that it's hard to make it worse.
> 
> I'm just at the end of a seven hour session of looking at these APIs and the related documentation and tickets.  I feel I'm gradually coming to understand what's going on but I still don't grasp the big picture (and I fear I still have a long way to go).  Whatever else you might want to say about this code, I don't think it's correct to say that what's going on here is "imposing a little cognitive burden".  It's imposing a lot more than that.

My point was not that these APIs are straightforward and easy to grasp.  I would have thought that my use of the adjective "crushing" communicated that :-).  It's clearly a complexity disaster and one I would very much like to address.  To be clear I understand that both TLS itself and these APIs _each_ impose a crushing cognitive burden, and I think you understand the TLS part already.

My point was that I don't feel like the one line of code you complained about - the private `ClientTLSOptions` import in the tests - is responsible for any significant portion of those seven hours.  If we had come up with a completely clean way to write this one test using only public APIs, my estimate was that this would have been a 6 hour and 50 minute exercise and that most of the problems here would remain.  So I would like to land this more or less as-is so that we can spend our time holistically analyzing the problem and making changes elsewhere that are unrelated to this fix.

Nevertheless, if you can't understand what is being done clearly enough to give it a proper code review, I do think that we are blocked until we can get it to that point.

**Comment by twisted-trac** [comment]
|[<img alt="hynek's avatar" src="https://avatars.githubusercontent.com/u/41240?s=50" width="50" height="50">](https://github.com/hynek)<a name="note_13"></a>|@hynek commented|
|-|-|

Since JP resigned from Twisted’s TLS on IRC the other day, I’ll try to move this thing forward now.

----


Before I go into this any deeper, let me state three axioms that the rest is building on:

1. Twisted’s TLS code is not optimal.  There’s cruft, weird APIs and leaky abstractions.  That, and the fact that Twisted’s TLS APIs changed a few times led to the unfortunate state of the complete lack of internal testing tools for TLS code.  I would go as far as saying that the current state is a big shit show with multiple generations of APIs happily co-existing (cf. [#6923](https://github.com/twisted/twisted/issues/6923) for instance) and chaos being all around.
2. *This* particular ticket is about a one line error that was an unfortunate oversight (OTOH, we pulled an Apple!).  And hilariously given the word count, I have nothing to say about the code (except maybe that the `called` flag is superfluous and if you insist on having it, the test method should be split up in two).
3. *All* criticism about this ticket was *solely* about how the *test* verifies that a method has been called with a certain argument.  This is not about public API design, this it not about correctness. 

Point 1 is a fact we can’t really change on short notice.  I hope one day there’ll be a `twisted.internet.tls` without all these warts.  But I’m afraid this is not that day.  And especially not as part of point 2 that is already in the wild as part of 14.0.2 as JP also noted.  Point 2 and 3 together do not make enough of an argument to explain why there was only one feature Twisted release in 2014 (albeit a very important one).


----


That said, we urgently need a serious effort to harmonize Twisted’s TLS code.  JP is completely right with his criticism and I kind of bemoaned it too a few times.  It’s just out of the scope of *this* ticket.

Part of that is [#6923](https://github.com/twisted/twisted/issues/6923) but also the addition of testing tools, verified fakes, transports and whatnot.  Also the explanation of how Agent interacts with TLS ought to be documented using an annotated graphics if it’s to be understood but mortal minds.  I remember re-understanding it several times when working on CVE-2014-2234.

We may collect these tasks and issues in https://twistedmatrix.com/trac/wiki/Plan/Security and I would offer my help with that if more *serious* helping hands *including* glyph would sign up to that.  This is not like my 2013/2014 stint where I iteratively fixed a few existing APIs; this is big picture stuff.

I also honestly feel the scope of this problem *cannot* be resolved on a ticket, wiki, or mailing list.  Hence I propose a formal meeting at PyCon in Montreal just about this complex of problems.

glyph’s proposals of a new TLS API for Twisted (which I largely agree with FWIW) is another thing that ought to be brought up.  Honestly I regret we spent so much time on discussing the design of the rather academic PyCa `tls` package instead of looking how to model a new Twisted TLS API (which would be a useful real-world pointer for the PyCA `tls` package too) last year.

----

Finally and with a heavy heart: 

- **glyph, please go on and merge.**
- **Hawkie, please go on and make 15.0.0 a thing.**


----


Please don’t make me regret this.  JP’s objections are serious problems I’ve encountered first-hand while working on Twisted’s TLS; they need to be addressed **ASAP**.  But ASAP realistically can’t mean this ticket.

P.S. I’m aware that I kind of picked up the TLS ball last year hence the fact, that this ticket blocked hawkie from releasing 14.1 for such a long time is also my fault and I’d like to offer my sincerest apologies to everyone.  But yeah, I’ve been freaking busy with other open source projects. :-/

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_14"></a>|@glyph set status to `closed`|
|-|-|

Resolved by r43662 - not sure why the commit hook didn't fire.  Perhaps it was just too sad about the state of things to proceed.

Let's definitely move forward to fix this though.  hynek, I look forward to working with you on it.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_15"></a>|@hawkowl commented|
|-|-|

Ticket retargeted after milestone deleted
